tensorflow/core/kernels/mkl_batch_matmul_op.cc

```diff
@@ -0,0 +1,195 @@
+    OP_REQUIRES_OK(context, context->GetAttr("adj_y", &adj_y_));
+  }
+
+  virtual ~BatchMatMulMkl() {}
+
+  void Compute(OpKernelContext *ctx) override {
+    const Tensor &in0 = ctx->input(0);
+    const Tensor &in1 = ctx->input(1);
+    OP_REQUIRES(ctx, in0.dims() == in1.dims(),
+                errors::InvalidArgument("In[0] and In[1] has different ndims: ",
+                                        in0.shape().DebugString(), " vs. ",
+                                        in1.shape().DebugString()));
+    const int ndims = in0.dims();
+    OP_REQUIRES(
+        ctx, ndims >= 2,
+        errors::InvalidArgument("In[0] and In[1] ndims must be >= 2: ", ndims));
+    TensorShape out_shape;
+    for (int i = 0; i < ndims - 2; ++i) {
+      OP_REQUIRES(ctx, in0.dim_size(i) == in1.dim_size(i),
+                  errors::InvalidArgument("In[0].dim(", i, ") and In[1].dim(",
+                                          i, ") must be the same: ",
+                                          in0.shape().DebugString(), " vs ",
+                                          in1.shape().DebugString()));
+      out_shape.AddDim(in0.dim_size(i));
+    }
+    auto n = (ndims == 2) ? 1 : out_shape.num_elements();
+    auto d0 = in0.dim_size(ndims - 2);
+    auto d1 = in0.dim_size(ndims - 1);
+    Tensor in0_reshaped;
+    CHECK(in0_reshaped.CopyFrom(in0, TensorShape({n, d0, d1})));
+    auto d2 = in1.dim_size(ndims - 2);
+    auto d3 = in1.dim_size(ndims - 1);
+    Tensor in1_reshaped;
+    CHECK(in1_reshaped.CopyFrom(in1, TensorShape({n, d2, d3})));
+    if (adj_x_) std::swap(d0, d1);
+    if (adj_y_) std::swap(d2, d3);
+    OP_REQUIRES(ctx, d1 == d2,
+                errors::InvalidArgument(
+                    "In[0] mismatch In[1] shape: ", d1, " vs. ", d2, ": ",
+                    in0.shape().DebugString(), " ", in1.shape().DebugString(),
+                    " ", adj_x_, " ", adj_y_));
+    out_shape.AddDim(d0);
+    out_shape.AddDim(d3);
+    Tensor *out = nullptr;
+    OP_REQUIRES_OK(ctx, ctx->allocate_output(0, out_shape, &out));
+    if (out->NumElements() == 0) {
+      return;
+    }
+    if (in0.NumElements() == 0 || in1.NumElements() == 0) {
+      functor::SetZeroFunctor<Device, Scalar> f;
+      f(ctx->eigen_device<Device>(), out->flat<Scalar>());
+      return;
+    }
+
+    const uint64 M = in0_reshaped.dim_size(adj_x_ ? 2 : 1);
+    const uint64 K = in0_reshaped.dim_size(adj_x_ ? 1 : 2);
+    const uint64 N = in1_reshaped.dim_size(adj_y_ ? 1 : 2);
+    auto in0_ptr = (in0.template flat<Scalar>().data());
+    auto in1_ptr = (in1.template flat<Scalar>().data());
+    auto out_ptr = (out->template flat<Scalar>().data());
+    std::vector<CBLAS_TRANSPOSE> transa_array;
+    std::vector<CBLAS_TRANSPOSE> transb_array;
+    std::vector<MKL_INT> m_array;
+    std::vector<MKL_INT> n_array;
+    std::vector<MKL_INT> k_array;
+    std::vector<Scalar> alpha_array;
+    std::vector<Scalar> beta_array;
+    std::vector<const Scalar *> a_array;
```

Comment:
For the non-constant array, please add 
```
a_array.reserve(n);
```
before the loop to avoid allocation overhead for large batch sizes.